### ISSUE TITLE
Link error when building fortran example in Visual Studio 2019 on Windows

### ISSUE THREAD

**Issue by xiangyue1993** [ddemidov/amgcl#153]
Hi,

I have some problem when building the fortran example using Visual Studio 2019 on Windows with the Intel fortran Compiler. The VS solution was generated by cmake from the amgcl source code, and link error occurred when I build the `possion project` with fortran code. The error description is as below:

```
amgcl.obj : error LNK2019: unresolved external symbol '_amgcl_params_setf' referenced in function '_AMGCL_mp_AMGCL_PARAMS_SETF' 
amgcl.obj : error LNK2019: unresolved external symbol '_amgcl_params_seti' referenced in function ' _AMGCL_mp_AMGCL_PARAMS_SETI'
amgcl.obj : error LNK2019: unresolved external symbol '_amgcl_params_sets' referenced in function ' _AMGCL_mp_AMGCL_PARAMS_SETS' 
poisson.obj :  error LNK2019: unresolved external symbol '_amgcl_params_create' referenced in function ' _MAIN__' 
poisson.obj :  error LNK2019: unresolved external symbol '_amgcl_solver_create_f' referenced in function ' _MAIN__' 
poisson.obj :  error LNK2019: unresolved external symbol '_amgcl_solver_report' referenced in function ' _MAIN__' 
poisson.obj :  error LNK2019: unresolved external symbol '__amgcl_solver_solve' referenced in function ' _MAIN__' 
poisson.obj :  error LNK2019: unresolved external symbol '_amgcl_solver_destroy' referenced in function ' _MAIN__' 
poisson.obj :  error LNK2019: unresolved external symbol '_amgcl_params_destroy' referenced in function ' _MAIN__' 
```
The other projects in this VS solution were all built successfully, including the project `libamgcl`. Then I checked the Linker settings of `possion project`, the static library `libamgcl` has been set as dependency. 

Do you know how to resolve this problem?

Thanks & Regards,

Xiang

**Comment by ddemidov** [comment]
Hi,

I don't have access to Fortran on Windows or Visual Studio. The poisson project does compile and link successfully on Linux, so I suspect there is a problem with symbol exports on Windows. Are you generating and building the libamgcl project with the provided CMakeLists.txt files? The project should include the [dll.def](https://github.com/ddemidov/amgcl/blob/master/lib/dll.def) file in that case, which lists the exported symbols. Same symbols are used in the [call_lib](https://github.com/ddemidov/amgcl/blob/master/examples/call_lib.cpp) C++ project, so I would also check if that project is linked successfully for you.

**Comment by ddemidov** [comment]
Looks like the linker expects a leading underscore in the function names. Could this be related: https://community.intel.com/t5/Intel-Fortran-Compiler/How-to-avoid-prepend-underscore-when-linking-a-fortran-lib-from/td-p/903194?

**Comment by ddemidov** [comment]
Can you see if replacing `__stdcall` with `__cdecl` [here](https://github.com/ddemidov/amgcl/blob/87689e023052cf0fc3c630a8e65d7cd262145656/lib/amgcl.h#L35) helps?

**Comment by xiangyue1993** [comment]
> Can you see if replacing `__stdcall` with `__cdecl` [here](https://github.com/ddemidov/amgcl/blob/87689e023052cf0fc3c630a8e65d7cd262145656/lib/amgcl.h#L35) helps?

Thanks for your reply. 

I replace the macro with __cdecl, and there was still link error, but only one unresolved symbol left:

```poisson.obj :  error LNK2019: unresolved external symbol '__amgcl_solver_solve' referenced in function ' _MAIN__' ```

**Comment by ddemidov** [comment]
The only thing I can see about `amgcl_solver_solve` that is different from the rest of the calls, is that it returns a struct:
https://github.com/ddemidov/amgcl/blob/87689e023052cf0fc3c630a8e65d7cd262145656/lib/amgcl.h#L117-L122

Not sure what to do here. Do you have any suggestions? I guess we could send the output struct as an output parameter into the call.

**Comment by ddemidov** [comment]
#154 is a possible fix. Can you please check if it works?

**Comment by xiangyue1993** [comment]
> #154 is a possible fix. Can you please check if it works?

I just make a test, now it works! Thank you very much!

**Comment by ddemidov** [comment]
Ok, I've merged #154 into master. Thanks for testing!
